### PR TITLE
flow generator

### PR DESCRIPTION
**Pull request by clewis7**
start incorporating flow generator 

current idea: 
- use pre-trained weights from ground truth
- train flow gen using `df.flow_generator.train()`
    -  have default model args, but can be user updated

### CODE REVIEW

**Comment by clewis7** [comment]
TODO/Ideas
- [ ] way to have dataset, dataloader based on entries in dataframe
    - for now will be passing in combined movies, in the future will need to do combination on the fly
- [ ] set default augs, how much should users have control over model params? (initial learning rate, batch size, mode)
- [ ] add final model params to dataframe
- [ ] model output columns as well with useful info, perhaps epoch losses 
- [ ] save model weights as checkpoint when training is finished to the same location as the dataframe
- [ ] save all other relevant output information in hdf5 file
- [ ] during training...tqdm to show progress, but also matplotlib to plot losses of each epoch? or just save losses to output to plot later (user might not care about)

File: animal_soup/viewers/__init__.py [hunk 1]
```diff
@@ -0,0 +1,13 @@
+from ._ethogram import EthogramVizContainer
+from ._ethogram_cleaner import EthogramCleaner
+from ._ethogram_comparison import EthogramComparison
```
Comment (kushalkolar) [inline animal_soup/viewers/__init__.py:3]: append Viz to each so we know it's a viz module, unify names

**Review by kushalkolar — CHANGES_REQUESTED** [review]

File: animal_soup/extensions/flow_gen_extensions.py [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+            raise ValueError(f"mode argument must be one of {TRAINING_OPTIONS.keys()}")
+
+        # check gpu_id
+        gpu_options = self._get_gpu_options()
+        if gpu_id not in gpu_options.keys():
+            raise ValueError(f"gpu_id: {gpu_id} not in {gpu_options}. "
+                             f"Please select a valid gpu.")
+
+        # reload weights from file, want to use pretrained weights
+        model = self._load_pretrained_model(weight_path=Path('/home/clewis7/repos/animal-soup/pretrained_checkpoints/flow_gen.ckpt'), mode="slow")
```
[the remark this answers is not in the document]
Comment (clewis7) [reply animal_soup/extensions/flow_gen_extensions.py:69]: yeah, Stefan and I talked about this some...I think there is no need to have the DEG model weights...only cj-model weights once we have trained with ground truth and use that as the default model weights

but then if someone wants to retrain, then the new model weights will get saved and those can become the default for that dataframe

need to think about how to make it easy for a user to use what checkpoint for the model weights, default should def be cj-weights, but then changing it after that

File: animal_soup/extensions/flow_gen_extensions.py [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+
+        # create available dataset from items in df
+        training_vids = list()
+        parent_data_path = get_parent_raw_data_path()
+        for ix, row in self._df.iterrows():
+            # for now assuming codec is AVI, but in future will need to detect or update
```
[the remark this answers is not in the document]
Comment (clewis7) [reply animal_soup/extensions/flow_gen_extensions.py:75]: yeah, had thought about this...either user needs to specify codec or need someway to detect what codec a video is...I think glob for all could be okay

File: animal_soup/extensions/flow_gen_extensions.py [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+            training_vids.append(parent_data_path.joinpath(row["animal_id"],
```
[the remark this answers is not in the document]
Comment (clewis7) [reply animal_soup/extensions/flow_gen_extensions.py:76]: hmm, I think it would be useful but then also make the dataframe a little messier visually....

File: animal_soup/extensions/flow_gen_extensions.py [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+        if "stop_method" not in STOP_METHODS.keys():
+            raise ValueError(f"stop_method: {stop_method} must be one of {STOP_METHODS}")
+        # if stop method == "learning rate", should be type float
+        if stop_method.keys()[0] == "learning_rate" and not isinstance(stop_method["learning_rate"], float):
+            raise ValueError("if stop_method is 'learning_rate', must be of type float")
+        # else should be type int
+        elif not isinstance(stop_method.values()[0], int):
+            raise ValueError("if 'stop_method' is either 'early' or 'num_epochs', must be type int")
+
+        # in notes column, add flow_gen_train params for model
```
[the remark this answers is not in the document]
Comment (clewis7) [reply animal_soup/extensions/flow_gen_extensions.py:91]: yeah, something like this...Stefan and I talked about how people in the lab will prob only want to specify the mode and perhaps something like batch size...but beyond that most thing should be default and then we could just expose them in a column

File: animal_soup/extensions/flow_gen_extensions.py [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+        # flow_gen output should also get stored in hdf5 file in same place as df path
+        # at end of training should also store new model checkpoint?
+
+
+        print("Starting training")
+        print(f"Training Mode: {mode} \n"
+              f"Model: {TRAINING_OPTIONS[mode]} \n"
+              f"Params: ")
+
+        return model
```
[the remark this answers is not in the document]
Comment (clewis7) [reply animal_soup/extensions/flow_gen_extensions.py:102]: yeah, I was just returning the model here for testing purposes to make sure the model had the weights from the checkpoint!! still very alpha lol

File: animal_soup/extensions/flow_gen_extensions.py [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+
+        # remove "model." preprend if exists
+        new_state_dict = OrderedDict()
+        for k, v in pretrained_model_state.items():
+            if k[:6] == 'model.':
+                name = k[6:]
+            else:
+                name = k
+            new_state_dict[name] = v
+        pretrained_model_state = new_state_dict
```
[the remark this answers is not in the document]
Comment (clewis7) [reply animal_soup/extensions/flow_gen_extensions.py:146]: some of the checkpoint have this weird 'model.' pre-pended to the layer weights and they  need to be the same as what the model you are trying to update the weights to will be...this is how DEG was doing it, but I am sure there is something better

File: animal_soup/__init__.py [hunk 8]
```diff
@@ -1,30 +1,10 @@
...
-    "DECORD_CONTEXT",
-    "EthogramComparison"
-]
+from .df_utils import (
+                    load_df,
+                    create_df,
+                    set_parent_raw_data_path,
+                    get_parent_raw_data_path
+                    )
+from .flow_generator import *
```
[the remark this answers is not in the document]
Comment (clewis7) [reply animal_soup/__init__.py:9]: yeah, was just a placeholder for type annotation where I was returning the model, but that is going to change

File: animal_soup/extensions/flow_gen_extensions.py [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+            Specify which gpu to use for training the model.
+        lr: float, default 0.0001
+            Default learning rate.
+        stop_method: Dict[str, Union[int, float]], default {num_epochs: 10}
+            Method for stopping training. Argument must be one of ["early", "learning_rate", "num_epochs"]
+
+            | stop method   | description                                                                |
+            |---------------|----------------------------------------------------------------------------|
+            | early         | Stop training if there is no improvement after a certain number of events  |
+            | learning_rate | Stop training when learning rate drops below a given threshold             |
```
[the remark this answers is not in the document]
Comment (clewis7) [reply animal_soup/extensions/flow_gen_extensions.py:54]: this is how DEG defines it 

```
"""Simple early stopper that stops when the learning rate drops below some threshold.
    Example usage: you reduce your learning rate when your validation loss stops improving. If the learning rate drops
        below some minimum value, automatically stop training.
```

File: animal_soup/extensions/flow_gen_extensions.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (clewis7) [reply animal_soup/extensions/flow_gen_extensions.py]: model in/model out kwargs

default is pretrained cj-weighs, unless otherwise specified by model Path kwargs

File: animal_soup/extensions/flow_gen_extensions.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (clewis7) [reply animal_soup/extensions/flow_gen_extensions.py:217]: use lstrip() instead

**Comment by clewis7** [comment]
closes #23

**Comment by clewis7** [comment]
THINGS LEFT TO DO FOR FLOW GEN:
- [ ] finish lightning module
    - [ ] finish implementing loss 
    - [ ] implement image reconstructor
    - [ ] finish implementing step process for model
- [ ] implement trainer using lightning module
- [ ] train DEG with ground truth data with each model to get checkpoints
- [ ] add column to dataframe for model params and save to disk

THINGS TO THINK ABOUT:
1. DEG does a lot of metric computations, what is sufficient for us? 
    - my current thoughts are to just calculate the batch loss through every epoch and at the end of each epoch plot the loss and output
    - DEG uses SSIM flow metric also for flow gen in addition to loss
2. What information should be being saved to output file? What is actually relevant to people in the lab? 
3. Should we save the checkpoint at every epoch for safety purposes? 

@kushalkolar thoughts?

File: animal_soup/flow_generator/train.py [hunk 10]
```diff
@@ -0,0 +1,283 @@
...
+    def apply_gpu_transforms(self, images: torch.Tensor) -> torch.Tensor:
+        """Apply the flow generator GPU transformations to a video."""
+        with torch.no_grad():
+            images = self.gpu_transforms(images).detach()
+        return images
+
+    def configure_optimizers(self):
+        """Configure the optimizer to be used in training the flow generator."""
+
+        weight_decay = 0  # if self.hparams.weight_decay is None else self.hparams.weight_decay
```
Comment (kushalkolar) [inline animal_soup/flow_generator/train.py:121]: what's the commented out code??

File: animal_soup/flow_generator/train.py [hunk 11]
```diff
@@ -0,0 +1,283 @@
...
+        Can be a string (ex: "simple", "advanced") or a Pytorch Lightning Profiler instance. Gives metrics
+        during training.
+
+    Returns
+    -------
+    trainer: pl.Trainer
+        A trainer to be used to manage training the flow generator.
+    """
+
+    tensorboard_logger = pl.loggers.tensorboard.TensorBoardLogger(
```
Comment (kushalkolar) [inline animal_soup/flow_generator/train.py:260]: run `black` on the codebase

File: animal_soup/utils/callbacks.py [hunk 12]
```diff
@@ -0,0 +1,120 @@
...
+              f"Train Loss: {metrics['loss']:.4f}")
+
+
+class PlotLossCallback(pl.Callback):
+    def __init__(self):
+        super().__init__()
+
+        self.total_losses = list()
+
+    def on_train_epoch_end(self, trainer: "pl.Trainer", pl_module: "pl.LightningModule") -> None:
```
Comment (kushalkolar) [inline animal_soup/utils/callbacks.py:28]: does this work as expected?

File: animal_soup/utils/transforms.py [hunk 13]
```diff
@@ -0,0 +1,343 @@
+"""Utility functions and classes for getting transformations or applying them."""
```
Comment (kushalkolar) [inline animal_soup/utils/transforms.py:1]: this is just directly taken from DEG right? Put a link.

File: setup.py [hunk 14]
```diff
@@ -15,8 +15,14 @@
     "tables",
     "jupyter-rfb",
     "jupyterlab",
+    "kornia",
+    "omegaconf",
```
Comment (kushalkolar) [inline setup.py:19]: put a comment for any "non obvious" dependencies, because we will forget why stupid ones like this are here when it's not used anywhere

**Review by kushalkolar — CHANGES_REQUESTED** [review]

File: animal_soup/data/datasets.py [hunk 15]
```diff
@@ -0,0 +1,213 @@
...
+from typing import *
+import numpy as np
+from pathlib import Path
+import os
+from vidio import VideoReader
+import random
+
+
+class SingleVideoDataset(data.Dataset):
+    """PyTorch Dataset for loading a set of sequential frames and applying pre-defined augmentations to each frame."""
```
[the remark this answers is not in the document]
Comment (clewis7) [reply animal_soup/data/datasets.py:12]: yes, one of the augmentations is jittering the color, which takes into account brightness, contrast, saturation, and hue

File: animal_soup/data/datasets.py [hunk 16]
```diff
@@ -0,0 +1,213 @@
...
+
+        self._zeros_image = None
+
+    def get_zeros_image(self, c, h, w):
+        """Zero image frames to be added to front or back of image stack."""
+        if self._zeros_image is None:
+            # ALWAYS ASSUME OUTPUT IS TRANSPOSED
+            self._zeros_image = np.zeros((c, h, w), dtype=np.uint8)
+            for i in range(3):
+                self._zeros_image[i, ...] = self.mean_by_channels[i]
```
[the remark this answers is not in the document]
Comment (clewis7) [reply animal_soup/data/datasets.py:73]: yeah, this is what gets calculated in the normalization augmentation and then you can use the mean by channel to normalize images across your dataset

File: animal_soup/data/datasets.py [hunk 17]
```diff
@@ -0,0 +1,213 @@
...
+        # if frames per clip is 11, dataset[0] would have 5 blank frames preceding, with the 6th-11th being real frames
+        blank_start_frames = max(self.frames_per_clip // 2 - index, 0)
+
+        framecount = self.metadata['framecount']
+
+        start_frame = index - self.frames_per_clip // 2 + blank_start_frames
+        blank_end_frames = max(index - framecount + self.frames_per_clip // 2 + 1, 0)
+        real_frames = self.frames_per_clip - blank_start_frames - blank_end_frames
+
+        seed = np.random.randint(2147483647)
```
[the remark this answers is not in the document]
Comment (clewis7) [reply animal_soup/data/datasets.py:141]: The append/prepend does not happen until you call `__get_item__`, and I think setting the seed makes sure things remain consistent if you call `__get_item__` more than once...not sure why they did this

File: animal_soup/extensions/flow_gen_extensions.py [hunk 18]
```diff
@@ -0,0 +1,332 @@
...
+    """
+    Pandas dataframe extensions for training the flow generator.
+    """
+
+    def __init__(self, df):
+        self._df = df
+
+    def train(self,
+              mode: str = "slow",
+              flow_window: int = 11,
```
[the remark this answers is not in the document]
Comment (clewis7) [reply animal_soup/extensions/flow_gen_extensions.py:51]: yeah, I think they use 11 as a general but because our task is so specific that maybe decreasing this would be okay

File: animal_soup/extensions/flow_gen_extensions.py [hunk 19]
```diff
@@ -0,0 +1,332 @@
...
+            generator.
+
+            | mode   | model           |
+            |--------|-----------------|
+            | fast   | TinyMotionNet   |
+            | medium | MotionNet       |
+            | slow   | TinyMotionNet3D |
+
+        flow_window: int, default 11
+            Window size for computing optical flow features of a frame. Recommended to be the minimum number of frames
```
[the remark this answers is not in the document]
Comment (clewis7) [reply animal_soup/extensions/flow_gen_extensions.py:75]: yes, but I did not implement it...I can though

File: animal_soup/extensions/flow_gen_extensions.py [hunk 20]
```diff
@@ -0,0 +1,332 @@
...
+            If you want to retrain the model using different model weights than the default. User can
+            provide a location to a different model checkpoint. For example, if you had retrained the flow generator
+            previously and wanted to use those weights instead.
+        model_out: str or Path, default None
+            User provided location of where to store model output such as model checkpoint with updated weights,
+            hdf5 file with model results/metrics, etc. Should be a directory. By default, the model output will get stored in the same
+            directory as the dataframe.
+        """
+        # check valid model_in if not None
+        if model_in is not None:
```
[the remark this answers is not in the document]
Comment (clewis7) [reply animal_soup/extensions/flow_gen_extensions.py:102]: that is because it gets parsed later in loading the model from a checkpoint if model in is `None`

File: animal_soup/utils/callbacks.py (hunk 12 above)
Comment (clewis7) [reply animal_soup/utils/callbacks.py:28]: yes!!

File: animal_soup/flow_generator/train.py (hunk 10 above)
Comment (clewis7) [reply animal_soup/flow_generator/train.py:121]: irrelevant, you can have weight decay as a param but then DEG just uses zero no matter what

File: animal_soup/flow_generator/train.py [hunk 21]
```diff
@@ -0,0 +1,283 @@
...
+
+        if isinstance(self.model, TinyMotionNet3D):
+            self.gpu_transforms = get_gpu_transforms(augs=self.augs, conv_mode='3d')['train']
+        else:
+            self.gpu_transforms = get_gpu_transforms(augs=self.augs)['train']
+
+        if model_in is None:
+            self.model_in = Path('/home/clewis7/repos/animal-soup/pretrained_models/flow_generator').joinpath(
+                MODEL_MAP[type(self.model)]).with_suffix('.ckpt')
+        else:  # no need to validate because model weights will have already been loaded
```
[the remark this answers is not in the document]
Comment (clewis7) [reply animal_soup/flow_generator/train.py:81]: that there is no need to validate the `model_in` path again here because it will have already been validated when loading in the model weights

File: animal_soup/flow_generator/train.py (hunk 10 above)
Comment (kushalkolar) [reply animal_soup/flow_generator/train.py]: I guess it's a carried over comment from DEG code? I'd just delete it because it's confusing like why is this there
